src/librustc_driver/driver.rs

```diff
@@ -1275,9 +1275,10 @@ pub fn build_output_filenames(input: &Input,
                                            .values()
                                            .filter(|a| a.is_none())
                                            .count();
-            let ofile = if unnamed_output_types > 1 {
-                sess.warn("ignoring specified output filename because multiple outputs were \
-                           requested");
+            let ofile = if unnamed_output_types > 1 &&
+                        !sess.opts.output_types.contains_key(&OutputType::Exe) {
```

Comment:
This causes the warning to only be printed if linking was `--emit=link` was not passed, which is the opposite of what's wanted.